### PR TITLE
Add waveform modes module

### PR DESCRIPTION
**Pull request by cdcapano**
This is part 1 of 3 patches to replace #3456. This patch adds a module to the waveform package that will generate waveforms one mode at a time, without applying spherical harmonics. The convention I've used is that these functions return the real and imaginary parts of the complex hlm time series as a tuple. For the FD version, it will return the equivalent of the FFT of the real and imaginary parts of the complex hlm time series. That way we do not need to worry about handling complex -> complex FFTs.

Currently, this only works with the NRSurrogate (and so only the `get_td_waveform_modes` function actually does anything). The problem with the Phenom waveforms is they combine the +/-m modes together. There is a way to separate those by calling from two different phase and/or inclination angles, but getting conventions right requires some work. I have the start of a function in there to do that for PhenomX. I've left it there for now. I think they're working on a unified convention for generating waveforms mode-by-mode in lalsimulation. In which case, I'll switch to calling that when it's released. If not, I'll update the PhenomX function at some point.

This also adds a couple of wrapper functions around the functions to convert J <-> L0 frame. I ended up not using that for this, but will be useful for other things (like to plot `thetaJN` rather than inclination), so I've left them in.

### CODE REVIEW

File: pycbc/waveform/waveform_modes.py
Comment (spxiwh) [inline pycbc/waveform/waveform_modes.py:71]: Looks a bit strange to use `ell` for l and `m` for m. I think `l` and `m` is fine here, but if you want to appease CodeClimate `l_mode` and `m_mode` might be usable?

File: pycbc/waveform/waveform_modes.py
Comment (spxiwh) [inline pycbc/waveform/waveform_modes.py:88]: Are these the modes in the co-precessing frame, or in the radiation frame? I assume that the radiation frame is what is really important given what's above ... But then `XPHM` *must* have the other 2,X modes in the radiation frame or it's just broken!

(I think this does need to be in the docstrings, but interesting what is intended with precessing waveforms first.)

Comment (spxiwh) [reply pycbc/waveform/waveform_modes.py:88]: Yeah ... Actually I think there is inconsistency here, as I think you have defined the Phenom modes in the co-precessing frame, but NRSur seems to be in radiation frame, as no NR waveform is modelling (4,0) mode [well]!

File: pycbc/waveform/waveform_modes.py
Comment (spxiwh) [inline pycbc/waveform/waveform_modes.py:106]: I can see what this is doing, but the docstring is confusing. Am I right in saying that this will always return a real number, and then this is the magnitude of the Y_lm?

File: pycbc/waveform/waveform_modes.py
Comment (spxiwh) [inline pycbc/waveform/waveform_modes.py:143]: I don't like that we should have to define this for every waveform model. I'm reasonably sure that `lalsimulation` defines a generic `get_specific_modes` function (although probably not all waveforms are plugged into it). Did this never actually happen, or is it not usable for some reason?

File: pycbc/waveform/waveform_modes.py
Comment (spxiwh) [inline pycbc/waveform/waveform_modes.py:157]: Just to check, does this have to copy the data in memory, or does it just change the pointers around? If the former, is the cost of this negligble?

File: pycbc/waveform/waveform_modes.py
Comment (spxiwh) [inline pycbc/waveform/waveform_modes.py:281]: If Phenom is somehow using something different (which is a problem, as lalsimulation does not allow for this freedom), how will the user know that this function won't work?

File: pycbc/waveform/waveform_modes.py
Comment (spxiwh) [inline pycbc/waveform/waveform_modes.py:476]: `real` -> `imaginary` in the docstring, I think.

**Review by spxiwh — CHANGES_REQUESTED** [review]
Thanks for splitting this up. It makes it much easier to see what is being added!

I left some comments/questions. Also see CodeClimate. I think the "Too Many XXX" comments can be ignored, using `l` and `m` for the modes probably makes sense (but not mixing `ell` and `m`), and `FIXME`s are appropriate but the other things should be resolved (including other "too short variable name" complaints).

... As an aside, why did you choose `ell`, the "official" spelling of the letter is `el` (https://en.wikipedia.org/wiki/L) (most of the letter names make great scrabble words).

File: pycbc/waveform/waveform_modes.py
Comment (ahnitz) [reply pycbc/waveform/waveform_modes.py:88]: They must be defined in the intertial frame (closely related to the radiation frame but remember there is not inclination at this point).

Comment (ahnitz) [reply pycbc/waveform/waveform_modes.py:88]: I'll let Collin respond, but my understanding is that *only* the NRSur models (or rather two of them) actually support this output at the moment.

File: pycbc/waveform/waveform_modes.py
Comment (ahnitz) [reply pycbc/waveform/waveform_modes.py:143]: There is no such infrastructure at the moment, though there appear to be some remnants of past efforts. In particular there are partial functions for XPHM and Pv3HM which either outright don't work and are disable, or use an inconsistent convention which is not possible to combine with spherical harmonics afterwards.

Comment (ahnitz) [reply pycbc/waveform/waveform_modes.py:143]: @spxiwh We've confirmed with Seb that we haven't missed something and have raised our support for a generic interface in the future, which would make it easier for us to pick up lalsimulation models. Of course, we'd like to be generic and allow for models outside lalsimulation, but it is indeed  a pain if they all use different conventions.

Comment (spxiwh) [reply pycbc/waveform/waveform_modes.py:143]: Fair enough.

File: pycbc/waveform/waveform_modes.py
Comment (spxiwh) [reply pycbc/waveform/waveform_modes.py:88]: Yeah, good point about radiation frame.

... I think that SpinTaylorTX waveforms also support this output, but that may not be relevant for you and could be added as/when needed if there's no consistent interface [yet] in lalsim.

File: pycbc/waveform/waveform_modes.py
Comment (cdcapano) [reply pycbc/waveform/waveform_modes.py]: It was habit from LaTeX. Switched to `l` (eat it, CodeClimate!).

File: pycbc/waveform/waveform_modes.py
Comment (cdcapano) [reply pycbc/waveform/waveform_modes.py:88]: Yeah, at the moment there's no common interface. @Cyberface had mentioned that they might have something in the works, but don't know when that will be.

I haven't used SpinTaylor in a looonngg time...

File: pycbc/waveform/waveform_modes.py
Comment (cdcapano) [reply pycbc/waveform/waveform_modes.py]: Yes, that's right. I've fleshed out the doc string.

File: pycbc/waveform/waveform_modes.py
Comment (cdcapano) [reply pycbc/waveform/waveform_modes.py:176]: Good question... I'm not sure. I remember being burned by not copying data returned by lalsimulation in the past, but that was several years ago; not sure if it's changed or not. Looking at what's done in `get_td_waveform` though (https://github.com/gwastro/pycbc/blob/master/pycbc/waveform/waveform.py#L214) the data is copied, so probably best to let it copy.

File: pycbc/waveform/waveform_modes.py
Comment (cdcapano) [reply pycbc/waveform/waveform_modes.py:300]: 🤷‍♂️ I'm still not certain on what the different conventions are for the different approximants. But the `J -> L0` frame conversion should be applicable to the frame that `ChooseFDWaveform` and `ChooseTDWaveform` use. That function is actually what they use in LALInference and Bilby to convert inclination into thetaJN. It's just that when you start digging down and trying to call individual approximant's functions that you get into the weeds of trying to figure out what convention a particular waveform uses. Anyway, I ended up not actually using these functions in this module. I decided to leave them in though since we currently don't have a function in pycbc that will do the conversion between inclination and thetaJN. In another PR I'll probably add a wrapper function specifically for that in `conversions.py`.

File: pycbc/waveform/waveform_modes.py
Comment (cdcapano) [reply pycbc/waveform/waveform_modes.py]: Whoops! Thanks, fixed.

**Comment by cdcapano** [comment]
@spxiwh I think I addressed all your comments.

File: pycbc/waveform/waveform_modes.py
Comment (spxiwh) [reply pycbc/waveform/waveform_modes.py:88]: So it sounds like only `NRSur7dq4` is actually good to go right now? I think based on the above, this *definitely* cannot do the right thing for the Phenom precessing approximants. I would prefer these be removed for now (I can believe that the non-precessing Phenom approximants are okay).

File: pycbc/waveform/waveform_modes.py
Comment (spxiwh) [reply pycbc/waveform/waveform_modes.py:300]: Okay, I'm happy with the `l0frame_to_jframe` function but should we really keep a `phenom_l0frame_to_jframe` function as well? My take is that if the waveform is wrong, it should be fixed!

**Comment by spxiwh** [comment]
CodeClimate still seems to be reporting a couple of genuine issues?

**Review by spxiwh — APPROVED** [review]
I'm giving this approval to avoid holding this up any longer, but please do address remaining concerns before merging.

File: pycbc/waveform/waveform_modes.py
Comment (cdcapano) [reply pycbc/waveform/waveform_modes.py:232]: Fair enough; removed.

File: pycbc/waveform/waveform_modes.py
Comment (cdcapano) [reply pycbc/waveform/waveform_modes.py:88]: This block of code is just returning what the default modes are for each approximant. That is correct, and is generally useful, so leaving in.

**Comment by cdcapano** [comment]
The rest of the code climate issues are due to the comment blocks. There is the unreachable code that you get from the `NotImplementedError`, but I left that in on purpose. I've made that function a private method and added a message to the doc string clarifying that it doesn't work. Will merge as soon as Travis passes.

File: pycbc/waveform/waveform_modes.py
Comment (spxiwh) [reply pycbc/waveform/waveform_modes.py:88]: I agree that it's returning the default modes for `IMRPhenomXHM` in the normal coordinate system that we are using and that this is useful. However, it is *not* returning the default modes for `IMRPhenomXPHM` in that same coordinate system (surely) and I'm worried about the potential for confusion (or mistakes) further down the line related to this.

**Pull request merged**
